server/src/server.ts

```diff
@@ -36,1440 +28,561 @@ import {
+		//  * spawn 2 fstar processes: one for typechecking, one lax process for fly-checking and symbol lookup
+		//  * set event handlers to read the output of the fstar processes
+		//  * send the current document to both processes to start typechecking
+		this.documents.onDidOpen(e => {
+			this.onOpenHandler(e.document);
+		});
 
-interface IdeCodeFragment {
-	"code-digest": string;
-	range: FStarRange;
-}
+		// Only keep settings for open documents
+		this.documents.onDidClose(e => {
+			this.killFStarProcessesForDocument(e.document);
+			// Clear all diagnostics for a document when it is closed
+			this.connection.sendDiagnostics({
+				uri: e.document.uri,
+				lax: true,
+				diagnostics: []
+			});
+			this.connection.sendDiagnostics({
+				uri: e.document.uri,
+				lax: false,
+				diagnostics: []
+			});
+		});
 
-interface IdeProgress {
-	stage: 'full-buffer-started'
-		| 'full-buffer-fragment-started'
-		| 'full-buffer-fragment-ok'
-		| 'full-buffer-fragment-lax-ok'
-		| 'full-buffer-fragment-failed'
-		| 'full-buffer-finished';
-	ranges: FStarRange;
-	"code-fragment"?: IdeCodeFragment
-}
 
+		// The content of a text document has changed. This event is emitted
+		// when the text document first opened or when its content has changed.
+		this.documents.onDidChangeContent(change => {
+			this.pendingChangeEvents.push(change.document);
+		});
 
-// An auto-complete response
-type IdeAutoCompleteResponse = [number, string, string];
-type IdeAutoCompleteResponses = IdeAutoCompleteResponse[];
-type IdeQueryResponseTypes = IdeSymbol | IdeError | IdeError[] | IdeAutoCompleteResponses;
+		this.documents.onDidSave(change => {
+			if (this.configurationSettings.verifyOnSave) {
+				this.pendingChangeEvents = []; //don't send any pending change events
+				this.validateFStarDocument(change.document, "full", false);
+				if (this.configurationSettings.flyCheck) {
+					this.validateFStarDocument(change.document, "lax", true, "lax"); //retain flycheck markers for the suffix
+				}
+			}
+		});
 
-// A query response envelope
-interface IdeQueryResponse {
-	'query-id': string;
-	kind: 'protocol-info' | 'response' | 'message';
-	status?: 'success' | 'failure';
-	level?: 'progress' | 'proof-state' | 'info';
-	response?: IdeQueryResponseTypes;
-	contents?: IdeProofState | IdeSymbol | IdeProgress;
-}
 
-type IdeResponse = IdeQueryResponse | ProtocolInfo
+		// Register connection handlers.
+		//
+		// Note: when passed as functions, `this` is not bound for class methods
+		// (and will therefore be undefined). We instead need to either use the
+		// method within a closure, or explicitly bind it here. See
+		// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/thiscallbacks
+		// for the official documentation on this behavior, and
+		// https://javascript.info/bind for a better explanation.
+		this.connection.conn.onInitialize(this.onInitialize.bind(this));
```

Comment:
```suggestion
		this.connection.conn.onInitialize(() => this.onInitialize());
```

The javascript semantics for `this` are really hairy (so hairy that you pasted the MDN explanation above).  Usually it's best to just use closures everywhere.